dask/bag/core.py

```diff
@@ -183,6 +183,42 @@ def finalize_item(results):
     return results[0]
 
 
+def merge_kwargs(dsk, kwargs):
+    """ Merges dask values from kwargs with dsk.
+
+    Returns the merged dsk and a graph expression useful for use with apply in
+    a dask graph to pass the final computed **kwargs to a function.
+
+    Currently only dask.bag.Item and python literal values are supported.
+    """
+    kw_pairs = []
+    for key, val in iteritems(kwargs):
+        if isinstance(val, Item):
+            dsk = merge(dsk, val.dask)
+            val = val.key
+        # TODO elif isinstance(val, Value):
+        elif isinstance(val, Base):
+            raise NotImplementedError(
+                '%s not supported as kwarg value to Bag.map_partitions'
+                % type(val).__name__)
+        kw_pairs.append([key, val])
+    return dsk, kw_pairs
```

Comment:
OK, that makes sense.  Thoughts on the following?

```python
def unpack_kwargs_for_map(kwargs):
    dsk = {}
    for k, v in kwargs.items():
        if hasattr(v, 'dask'):
            if hasattr(v, 'key'):
                kwargs[k] = v.key
                dsk.update(v.dask)
            else:
                raise ValueError(...)
    return dsk, kwargs
```
    
This has the benefit of supporting both `Value`s and `Item`s and it's also pure and so easily testable.

Unfortunately, this still lacks support for any sort of nesting, so if someone does the following then they'll get unfriendly behavior.

    b.map(func, keyword=[b.sum()])

This nesting case is where something more general could really be handy.  I agree with you though, that the fact that we want to limit ourselves by the contained dask collections makes things trickier.

I wonder if @jcrist has a suggestion.